### DISCUSSION TITLE
Should `dynamoose.type.CONSTANT(value)` set the default attribute?

### DISCUSSION THREAD

**Discussion by ptejada** [dynamoose/dynamoose#1492] [Q&A] [answered]
I originally started configuring the Constant type the manual a long way:

```
    entityType: {
      default: 'User',
      type: {
        value: 'Constant',
        settings: {
          value: 'User'
        }
      }
    },
```
Then I found out about the [dynamoose.type.CONSTANT](https://dynamoosejs.com/guide/Dynamoose#dynamoosetypeconstantvalue) helper but it doesn't set the `default` attribute and I don't know why. 

Is there any use case for the `Constant` attribute type without the `default` attribute?

**Comment by fishcharlie** [comment, marked as answer]
@ptejada Wouldn't a use case for this be having two different schemas in 1 model, or two models in 1 table. With similar attributes, but an `entityType` as you mentioned. In that case, it is plausible that the application (or caller) would need to define the `entityType` because the attributes are similar enough and Dynamoose doesn't know which one to pick.

**Reply by ptejada** [reply]
You make a valid point. I thought the purpose of the Constant attribute type was to more explicitly identify schemas so that as long as you provide `entityType` the right schema would be selected.

Either way, my impression so far is that the current state of the lib carries a lot of baggage from previous versions. I personally just got started with v3. For example, I am assuming the multi-schema option for models has to be an older functionality. Otherwise, it doesn't make much sense to have a multi-schema model when you can have each entity be its own model for the same table.

At least in the context of having one table and many single schema models, it can be useful to have the option to set the `default` as well. Mabe a second boolean param `dynamoose.type.CONSTANT(type: string, default: bool)`

**Reply by fishcharlie** [reply]
> For example, I am assuming the multi-schema option for models has to be an older functionality. Otherwise, it doesn't make much sense to have a multi-schema model when you can have each entity be its own model for the same table.

Correct. Multiple schemas in a single model was introduced in v2. Back when a model represented a DynamoDB table. In v3 a model represents a specific entity.

There might still be use cases for multiple schemas tho. For example maybe you have an entity like `Asset`. You could define multiple schemas for each asset type, that have individual properties. For example:

```js
const Asset = dynamoose.model("Asset", [
	{
		"id": String,
		"url": String,
		"type": dynamoose.TYPE.CONSTANT("asset"),
		"assetType": dynamoose.TYPE.CONSTANT("image"),
		"width": Number,
		"height": Number
	},
	{
		"id": String,
		"url": String,
		"type": dynamoose.TYPE.CONSTANT("asset"),
		"assetType": dynamoose.TYPE.CONSTANT("video"),
		"duration": Number
	}
]);
```

I'm assuming this will be fleshed out more overtime. Right now, I don't think there is any advantage to doing it this way vs breaking those out into their own models. But possibly, that could change. Feedback like this is helpful to shaping the future of Dynamoose and figuring out the direction we want to go.

One more note on this. Breaking changes are tough. In early stages of v3, it was required that you define a `Table` instance for your Model even if you never used it. This added friction to the upgrade process from v2. However, it made it more clear to users that a Model no longer represents a DynamoDB Table. In a late beta version of v3, I decided to remove that requirement, and say that when you first use a Model, and a Table hasn't been associated with it, it'll create one automatically for you. This made the upgrade process a lot easier for users.

For some historical context on this v1 -> v2 was a **very** tough upgrade due to the fact that it was a complete rewrite from the ground up. There is a tradeoff between making breaking changes, and having an easy upgrade path. And I'm still working to figure out a happy balance between those two items. v2 was rough for a lot of users tho.

I've really appreciated the increased engagement across the community since the launch of v3. I hope that continues into the future, as we work together to define these things, and figure out what is best.

> it can be useful to have the option to set the default as well. Mabe a second boolean param `dynamoose.type.CONSTANT(type: string, default: bool)`

I like this idea. However, I'm thinking about the implementation side of things. Right now `dynamoose.type.CONSTANT` just takes in the parameter, and returns an object that looks like the following:

```js
{
	"value": "Constant",
	"settings": {
		value
	}
}
```

So you can either pass it in like:

```js
{
	"id": String,
	"type": dynamoose.TYPE.CONSTANT("asset")
}
```

OR:

```js
{
	"id": String,
	"type": {
		"type": dynamoose.TYPE.CONSTANT("asset")
	}
}
```

In the first variation, it's pretty easy to simply add a `default` property to the returned object. However, in the second variation, the `default` property you pass in, won't be in the expected format. Which means we'd have to check for the default property both within the property option itself, as well as the `type` property. Which again, gets back to your point about baggage, that would be adding a lot of baggage, and unlike previous decisions, I'm not sure the pros outweigh the cons.

As in, what happens if you do the following?

```js
{
	"id": String,
	"type": {
		"type": dynamoose.TYPE.CONSTANT("asset", false),
		"default": "asset"
	}
}
```

Or to be more explicit:

```js
{
	"id": String,
	"type": {
		"type": {
			"value": "Constant",
			"settings": {
				"value": "asset",
				"default": false
			}
		},
		"default": "asset"
	}
}
```


In that case you **explicitly** setting the default value to `false` in the constant type, which is more explicit than `undefined`. Yet, you are also setting a `default` property. Which option should Dynamoose respect? Should it use the default or not?

Of course this is an edge case, and hopefully users won't do this. But it does fork a property and give it two different meanings, in two completely different places. We do similar things in other places of the library, however I'd argue that those cases are alias names, and all exist on the same root object. It's not embedding it within a completely different object.

Another thing I strive to do here, is ask myself: "How would I justify this in the documentation?" or "How would I explain the value or need of this to a user?". Alias options are pretty easy to explain: "It gives you another naming convention to use without any functional differences". This is different, the only valid justification here I can think of is "It was designed to work around an edge case". That is not a good reason in my opinion, and is why I'd lean towards not adding this.

---

The ideal scenario here would be to maybe figure out a way for the `dynamoose.TYPE.CONSTANT` value to break out of it's shell instead of just returning an object, and instead set the actual default parameter itself, to keep it in one place. That way even if you do:

```js
{
	"id": String,
	"type": {
		"type": dynamoose.TYPE.CONSTANT("asset", true)
	}
}
```

The resulting Schema that Dynamoose reads would be:

```js
{
	"id": String,
	"type": {
		"type": {
			"value": "Constant",
			"settings": {
				"value": "asset"
			}
		},
		"default": "asset"
	}
}
```

That way we keep things consistent, and it's not like we are forking a property. What that involves would likely be modifying the Schema ourself on initialization. Which is doable, and probably the way we'd want to go. But it isn't the most straight forward change in the world.

Feel free to submit a feature request for this.

---

Hopefully this helps.

**Reply by ptejada** [reply]
Thank you for the detailed response. In my case, this version will be good enough
```js
    entiyType: {
      default: type,
      type: dynamoose.type.CONSTANT(type)
    },
```
I understand finding the balance between flexibility and backward compatibility can be challenging. However, often times this can lead to a project that is difficult to maintain with all the different scenarios that need to be considered. It would be good to include an example section in the documentation to showcase best practices for common scenarios.
